src/gepa/core/state.py

```diff
@@ -106,68 +119,127 @@ def load(run_dir: str) -> "GEPAState":
 
         assert len(state.program_candidates) == len(state.parent_program_for_candidate)
         assert len(state.program_candidates) == len(state.named_predictor_id_to_update_next_for_program_candidate)
-
         return state
 
+    @staticmethod
+    def _migrate_from_legacy_state_v0(d: dict[str, Any]) -> None:
+        legacy_scores: list[list[float]] = d.pop("prog_candidate_val_subscores", [])
+        # convert to sparse val subscores
+        d["prog_candidate_val_subscores"] = [
+            {idx: score for idx, score in enumerate(scores)} for scores in legacy_scores
+        ]
+
+        pareto_front = d.get("pareto_front_valset")
+        if isinstance(pareto_front, list):
+            d["pareto_front_valset"] = {idx: score for idx, score in enumerate(pareto_front)}
+
+        program_at_front = d.get("program_at_pareto_front_valset")
+        if isinstance(program_at_front, list):
+            d["program_at_pareto_front_valset"] = {idx: set(front) for idx, front in enumerate(program_at_front)}
+
+        best_outputs = d.get("best_outputs_valset")
+        if isinstance(best_outputs, list):
+            d["best_outputs_valset"] = {idx: list(outputs) for idx, outputs in enumerate(best_outputs)}
+
+        d["validation_schema_version"] = GEPAState._VALIDATION_SCHEMA_VERSION
+
+    def get_program_average_val_subset(self, program_idx: int) -> tuple[float, int]:
+        scores = self.prog_candidate_val_subscores[program_idx]
+        if not scores:
+            return float("-inf"), 0
+        num_samples = len(scores)
+        avg = sum(scores.values()) / num_samples
+        return avg, num_samples
+
+    @property
+    def valset_evaluations(self) -> dict[DataId, list[ProgramIdx]]:
+        """
+        Valset examples by id and programs that have evaluated them. Keys include only validation
+        ids that have been scored at least once.
+        """
+        result = defaultdict(list)
+        for program_idx, val_scores in enumerate(self.prog_candidate_val_subscores):
+            for val_id in val_scores.keys():
+                result[val_id].append(program_idx)
+        return result
+
+    @property
+    def program_full_scores_val_set(self) -> list[float]:
+        return [
+            self.get_program_average_val_subset(program_idx)[0]
+            for program_idx in range(len(self.prog_candidate_val_subscores))
+        ]
+
+    def _update_pareto_front_for_val_id(
+        self,
+        val_id: DataId,
+        score: float,
+        program_idx: ProgramIdx,
+        outputs: dict[DataId, RolloutOutput] | None,
```

Comment:
Given that this function is defined only for one val_id, the outputs should be just output: RolloutOutput?